### PR TITLE
`rpc` package tests

### PR DESCRIPTION
**Pull request by coryschwartz**
Separated out the rpctest component from https://github.com/testground/testground/pull/1009 since there was question about the usefulness of the docker test and this part is unrelated.

Fix: https://github.com/testground/testground/issues/1004

### CODE REVIEW

File: pkg/rpc/rpc_test.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+			t.Error(err)
+			t.FailNow()
+		}
+		var ch rpc.Chunk
+		err = json.Unmarshal(body, &ch)
+		if err != nil {
+			t.Error(err)
+			t.FailNow()
+		}
+		t.Run("expected chunktype: "+test.f, func(t *testing.T) {
```
Comment (raulk) [inline pkg/rpc/rpc_test.go:60]: This usage of `t.Run` is weird. `t.Run` is used to run subtests, here you're using it to group asserts on test logic that has already run. Try grouping the entire for block inside `t.Run`?

File: pkg/rpc/rpc_test.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		t.Run("expected content: "+test.f, func(t *testing.T) {
+			p := ch.Payload.(string)
+			teststr := p
+			expected := test.expected
+			if test.ct == rpc.ChunkTypeProgress {
+				// ChunkTypeProgress contains unpredictable timestamps.
+				// Rather than deserializing, just check the last part of the data.
+				part := len(p) - 50
+				teststr = p[part:]
+				expected = test.expected[part:]
```
Comment (raulk) [inline pkg/rpc/rpc_test.go:76]: Can't we just use https://golang.org/pkg/bytes/#Contains?

Comment (raulk) [reply pkg/rpc/rpc_test.go:76]: or https://golang.org/pkg/strings/#Contains
or https://golang.org/pkg/strings/#HasSuffix

File: pkg/rpc/rpctest/rpctest.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package rpctest
```
[the remark this answers is not in the document]
Comment (coryschwartz) [reply pkg/rpc/rpctest/rpctest.go:1]: The reason for the separate package is that OutputWriter is used everywhere and I wanted to be able to import something in other packages. I originally wrote this so I could test docker packages in https://github.com/testground/testground/pull/1009 and just needed some kind of OutputWriter without using the http server to create it. But yeah, if there's a better way, lets do that instead.

File: pkg/rpc/rpc_test.go (hunk 1 above)
Comment (coryschwartz) [reply pkg/rpc/rpc_test.go:60]: Sure. I was just doing this for the output formatting.

File: pkg/rpc/rpc_test.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	// Fatal causes a crash. Just check the ones that send to the client.
+
+	tcs := []tc{
+		{"Info", rpc.ChunkTypeProgress, "TWF5IDE5IDA5OjUxOjQzLjAyMjI3OQkbWzM0bUlORk8bWzBtCXRlc3QJeyJyZXFfaWQiOiAiVGVzdExvZ2dpbmctSW5mbyJ9Cg=="},
+		{"Infof", rpc.ChunkTypeProgress, "TWF5IDE5IDA5OjUxOjQzLjAyMjUzNwkbWzM0bUlORk8bWzBtCXRlc3QJeyJyZXFfaWQiOiAiVGVzdExvZ2dpbmctSW5mb2YifQo="},
+		{"Infow", rpc.ChunkTypeProgress, "TWF5IDE5IDA5OjUxOjQzLjAyMjY1MAkbWzM0bUlORk8bWzBtCXRlc3QJeyJyZXFfaWQiOiAiVGVzdExvZ2dpbmctSW5mb3cifQo="},
+		{"Error", rpc.ChunkTypeProgress, "TWF5IDE5IDA5OjUxOjQzLjAyMjc1MQkbWzMxbUVSUk9SG1swbQl0ZXN0CXsicmVxX2lkIjogIlRlc3RMb2dnaW5nLUVycm9yIn0K"},
+		{"Errorf", rpc.ChunkTypeProgress, "TWF5IDE5IDA5OjUxOjQzLjAyMjg1MwkbWzMxbUVSUk9SG1swbQl0ZXN0CXsicmVxX2lkIjogIlRlc3RMb2dnaW5nLUVycm9yZiJ9Cg=="},
+		{"Errorw", rpc.ChunkTypeProgress, "TWF5IDE5IDA5OjUxOjQzLjAyMjk1NQkbWzMxbUVSUk9SG1swbQl0ZXN0CXsicmVxX2lkIjogIlRlc3RMb2dnaW5nLUVycm9ydyJ9Cg=="},
+		{"WriteResult", rpc.ChunkTypeResult, "test"},
```
[the remark this answers is not in the document]
Comment (coryschwartz) [reply pkg/rpc/rpc_test.go:32]: I'll take care of this today.

WriteError is a Json stream not just a single document
WriteBinary writes a []uint8

File: pkg/rpc/rpc_test.go (hunk 2 above)
Comment (coryschwartz) [reply pkg/rpc/rpc_test.go:76]: Yeah, I might re-think this though to not use just stiring comparisons,let me hack on this today.

File: pkg/rpc/rpctest/rpctest.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (raulk) [reply pkg/rpc/rpctest/rpctest.go:1]: Yes, but the tests are already in a separate package (`rpc_test`). This PR introduces:

* `pkg/rpc/rpc_test.go` under package `rpc_test`.
* `pkg/rpc/rpctest/rpctest.go` under package `rpctest`.

I think the latter is redundant.

File: pkg/rpc/rpctest/rpctest.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+	"github.com/testground/testground/pkg/rpc"
+)
+
+// NewRecordedOutputWriter returns an OutputWriter where the response is recorded.
+func NewRecordedOutputWriter(req_id string) (rec *httptest.ResponseRecorder, ow *rpc.OutputWriter) {
+	req := httptest.NewRequest("GET", "/", strings.NewReader(""))
+	req.Header.Add("X-Request-ID", req_id)
+	rec = httptest.NewRecorder()
+	ow = rpc.NewOutputWriter(rec, req)
+	return rec, ow
```
Comment (raulk) [inline pkg/rpc/rpctest/rpctest.go:16]: @coryschwartz ^^ still needs fixing. (golang.org/doc/effective_go.html#mixed-caps)

```suggestion
func NewRecordedOutputWriter(reqID string) (rec *httptest.ResponseRecorder, ow *rpc.OutputWriter) {
	req := httptest.NewRequest("GET", "/", strings.NewReader(""))
	req.Header.Add("X-Request-ID", reqID)
	rec = httptest.NewRecorder()
	ow = rpc.NewOutputWriter(rec, req)
	return rec, ow
```

**Pull request merged**